tools/extra/qpafilter/qpafilter.py

```diff
@@ -102,18 +106,74 @@ def write_sensor(self, sensor):
         upper_warning_id = self.threshold_map['Upper Warning']
         upper_fatal_id = self.threshold_map['Upper Fatal']
 
-        # TODO: endian-ness?
-        fmt = '<LLL'
-        self.outfile.write(struct.pack(fmt,
+        self.outfile.write(struct.pack(BLOB_STRUCT_FORMAT,
                                        sensor_id,
                                        upper_warning_id,
                                        warning))
-        self.outfile.write(struct.pack(fmt,
+        self.outfile.write(struct.pack(BLOB_STRUCT_FORMAT,
                                        sensor_id,
                                        upper_fatal_id,
                                        fatal))
 
 
+class blob_reader:
+    """Given an input file pointer, a sensor map, and a threshold map,
+       provide an iterator interface to reading the raw sensor data
+       items sequentially."""
+    def __init__(self, blobfp, sensor_map, threshold_map):
+        self.blob = blobfp.read()
+        self.start_marker_index = self.blob.find(BLOB_START_MARKER)
+        self.end_marker_index = self.blob.find(BLOB_END_MARKER)
+        self.sensor_map = sensor_map
+        self.threshold_map = threshold_map
+
+    class Iterator:
+        """Iterate over an in-memory blob. Calculate the number of
+           sensor entries by subtracting the offset of the first entry
+           from the offset of the end marker. Iterate and decode each
+           sensor entry, changing the raw data to human-readable output."""
+        def __init__(self,
+                     blob,
+                     start_marker_index,
+                     end_marker_index,
+                     sensor_map,
+                     threshold_map):
+            self.blob = blob
+            self.start_marker_index = start_marker_index
+            self.first_entry_index = self.start_marker_index + BLOB_STRUCT_SIZE
+            self.end_marker_index = end_marker_index
+            self.num_entries = ((self.end_marker_index -
+                                 self.first_entry_index) / BLOB_STRUCT_SIZE)
+            self.entry = 0
+            self.struct_iter = struct.iter_unpack(
+                BLOB_STRUCT_FORMAT,
+                self.blob[self.start_marker_index + len(BLOB_START_MARKER):])
+            self.sensor_map = sensor_map
+            self.threshold_map = threshold_map
+
+        def __next__(self):
+            if self.entry == self.num_entries:
+                raise StopIteration
+            self.entry += 1
```

Comment:
Do we use this?

Reply:
Yeah, line 155.  It's used to check when to stop iterating.

Reply:
Ah, we can also see if (sensor_id + threshold_id + raw_value) == BLOB_END_MARKER after line line 158 instead of relying on num_entries being a multiple of 12

Reply:
Hmm, BLOB_END_MARKER is 12 bytes now; but I'm not sure what it might end up being. The BMC team might want to use something of a different size.

Reply:
I meant that implicitly by the calculation of num_entries.